File: compat/suites/node-js-sdk/src/groups/cloudwatch-logs.ts
/**
 * groups/cloudwatch-logs.ts — CloudWatch Logs test groups for Node.js suite.
 *
 * Groups:
 *   logs-groups  — log group lifecycle (implemented)
 *   logs-streams — log stream lifecycle (implemented)
 *   logs-events  — PutLogEvents, GetLogEvents, FilterLogEvents (implemented)
 */

import {
  CreateLogGroupCommand,
  DeleteLogGroupCommand,
  DescribeLogGroupsCommand,
  CreateLogStreamCommand,
  DeleteLogStreamCommand,
  DescribeLogStreamsCommand,
  PutLogEventsCommand,
  GetLogEventsCommand,
  FilterLogEventsCommand,
  PutRetentionPolicyCommand,
  DeleteRetentionPolicyCommand,
  TagLogGroupCommand,
  ListTagsLogGroupCommand,
} from "@aws-sdk/client-cloudwatch-logs";
import { makeClients } from "../lib/clients.js";
import type { TestGroup } from "../lib/harness.js";
import * as assert from "node:assert/strict";

export function makeCloudWatchLogsGroups(suite: string): TestGroup[] {
  return [
    // ── logs-groups ────────────────────────────────────────────────────────
    {
      suite,
      service: "cloudwatch-logs",
      name: "logs-groups",
      tests: [
        {
          name: "CreateLogGroup",
          fn: async (ctx) => {
            const { logs } = makeClients(ctx);
            const groupName = `/overcast/${ctx.runId}/test`;
            // Should not throw — 200/no body on success
            await logs.send(
              new CreateLogGroupCommand({
                logGroupName: groupName,
              }),
            );
            const resp = await logs.send(
              new DescribeLogGroupsCommand({
                logGroupNamePrefix: groupName,
              }),
            );
            assert.ok(
              resp.logGroups?.some((g) => g.logGroupName === groupName),
              `CreateLogGroup: ${groupName} not found after create`,
            );
          },
        },
        {
          name: "DescribeLogGroups",
          fn: async (ctx) => {
            const { logs } = makeClients(ctx);
            const groupName = `/overcast/${ctx.runId}/test`;
            const resp = await logs.send(
              new DescribeLogGroupsCommand({
                logGroupNamePrefix: `/overcast/${ctx.runId}`,
              }),
            );
            assert.ok(
              resp.logGroups?.some((g) => g.logGroupName === groupName),
              `DescribeLogGroups: ${groupName} not found`,
            );
          },
        },
        {
          name: "PutRetentionPolicy",
          fn: async (ctx) => {
            const { logs } = makeClients(ctx);
            await logs.send(
              new PutRetentionPolicyCommand({
                logGroupName: `/overcast/${ctx.runId}/test`,
                retentionInDays: 7,
              }),
            );
            const resp = await logs.send(
              new DescribeLogGroupsCommand({
                logGroupNamePrefix: `/overcast/${ctx.runId}/test`,
              }),
            );
            const group = resp.logGroups?.find(
              (g) => g.logGroupName === `/overcast/${ctx.runId}/test`,
            );
            assert.strictEqual(
              group?.retentionInDays,
              7,
              `PutRetentionPolicy: expected retentionInDays=7, got ${group?.retentionInDays}`,
            );
          },
        },
        {
          name: "VerifyRetentionPolicy",
          op: false,
          fn: async (ctx) => {
            const { logs } = makeClients(ctx);
            const groupName = `/overcast/${ctx.runId}/test`;
            const resp = await logs.send(
              new DescribeLogGroupsCommand({ logGroupNamePrefix: groupName }),
            );
            const group = resp.logGroups?.find(
              (g) => g.logGroupName === groupName,
            );
            assert.strictEqual(
              group?.retentionInDays,
              7,
              `VerifyRetentionPolicy: expected retentionInDays=7, got ${group?.retentionInDays}`,
            );
          },
        },
        {
          name: "DeleteRetentionPolicy",
          fn: async (ctx) => {
            const { logs } = makeClients(ctx);
            await logs.send(
              new DeleteRetentionPolicyCommand({
                logGroupName: `/overcast/${ctx.runId}/test`,
              }),
            );
            const resp = await logs.send(
              new DescribeLogGroupsCommand({
                logGroupNamePrefix: `/overcast/${ctx.runId}/test`,
              }),
            );
            const group = resp.logGroups?.find(
              (g) => g.logGroupName === `/overcast/${ctx.runId}/test`,
            );
            assert.strictEqual(
              group?.retentionInDays,
              undefined,
              "DeleteRetentionPolicy: retentionInDays still set",
            );
          },
        },
        {
          name: "CreateLogStream",
          fn: async (ctx) => {
            const { logs } = makeClients(ctx);
            const groupName = `/overcast/${ctx.runId}/test`;
            await logs.send(
              new CreateLogStreamCommand({
                logGroupName: groupName,
                logStreamName: "compat-stream-1",
              }),
            );
            const resp = await logs.send(
              new DescribeLogStreamsCommand({ logGroupName: groupName }),
            );
            assert.ok(
              resp.logStreams?.some(
                (s) => s.logStreamName === "compat-stream-1",
              ),
              "CreateLogStream: stream not found after create",
            );
          },
        },
        {
          name: "TagLogGroup",
          fn: async (ctx) => {
            const { logs } = makeClients(ctx);
            const groupName = `/overcast/${ctx.runId}/test`;
            await logs.send(
              new TagLogGroupCommand({
                logGroupName: groupName,
                tags: { env: "compat", run: ctx.runId },
              }),
            );
            const resp = await logs.send(
              new ListTagsLogGroupCommand({ logGroupName: groupName }),
            );
            assert.strictEqual(
              resp.tags?.env,
              "compat",
              `TagLogGroup: expected tag env=compat, got ${resp.tags?.env}`,
            );
          },
        },
        {
          name: "DeleteLogGroup",
          fn: async (ctx) => {
            const { logs } = makeClients(ctx);
            const groupName = `/overcast/${ctx.runId}/test`;
            await logs.send(
              new DeleteLogGroupCommand({ logGroupName: groupName }),
            );
            const resp = await logs.send(
              new DescribeLogGroupsCommand({ logGroupNamePrefix: groupName }),
            );
            assert.notStrictEqual(
              resp.logGroups?.some((g) => g.logGroupName, groupName),
              `DeleteLogGroup: ${groupName} still present`,
            );
          },
        },
      ],
      teardown: async (ctx) => {
        const { logs } = makeClients(ctx);
        try {
          await logs.send(
            new DeleteLogGroupCommand({
              logGroupName: `/overcast/${ctx.runId}/test`,
            }),
          );
        } catch {}
      },
    },

    // ── logs-events ────────────────────────────────────────────────────────
    {
      suite,
      service: "cloudwatch-logs",
      name: "logs-events",
      tests: [
        {
          name: "PutLogEvents",
          fn: async (ctx) => {
            const { logs } = makeClients(ctx);
            const groupName = `/overcast/${ctx.runId}/events`;
            const streamName = "stream-1";
            const resp = await logs.send(
              new PutLogEventsCommand({
                logGroupName: groupName,
                logStreamName: streamName,
                logEvents: [
                  { timestamp: Date.now() - 2000, message: "first log event" },
                  { timestamp: Date.now() - 1000, message: "second log event" },
                  {
                    timestamp: Date.now(),
                    message: JSON.stringify({
                      level: "info",
                      msg: "structured",
                    }),
                  },
                ],
              }),
            );
            assert.ok(
              !resp.rejectedLogEventsInfo,
              "PutLogEvents: events rejected as too old",
            );
          },
        },
        {
          name: "GetLogEvents",
          fn: async (ctx) => {
            const { logs } = makeClients(ctx);
            const groupName = `/overcast/${ctx.runId}/events`;
            const streamName = "stream-1";
            const resp = await logs.send(
              new GetLogEventsCommand({
                logGroupName: groupName,
                logStreamName: streamName,
              }),
            );
            assert.ok(
              (resp.events?.length ?? 0) >= 3,
              `GetLogEvents: expected >=3 events, got ${resp.events?.length}`,
            );
          },
        },
        {
          name: "FilterLogEvents",
          fn: async (ctx) => {
            const { logs } = makeClients(ctx);
            const groupName = `/overcast/${ctx.runId}/events`;
            const resp = await logs.send(
              new FilterLogEventsCommand({
                logGroupName: groupName,
                filterPattern: "structured",
              }),
            );
            assert.notStrictEqual(
              resp.events?.length ?? 0,
              0,
              "FilterLogEvents: expected at least one matching event",
            );
          },
        },
        {
          name: "DescribeLogStreams",
          fn: async (ctx) => {
            const { logs } = makeClients(ctx);
            const groupName = `/overcast/${ctx.runId}/events`;
            const resp = await logs.send(
              new DescribeLogStreamsCommand({ logGroupName: groupName }),
            );
            assert.ok(
              resp.logStreams?.some((s) => s.logStreamName === "stream-1"),
              "DescribeLogStreams: stream-1 not found",
            );
          },
        },
        {
          name: "DeleteLogStream",
          fn: async (ctx) => {
            const { logs } = makeClients(ctx);
            const groupName = `/overcast/${ctx.runId}/events`;
            await logs.send(
              new DeleteLogStreamCommand({
                logGroupName: groupName,
                logStreamName: "stream-1",
              }),
            );
            const resp = await logs.send(
              new DescribeLogStreamsCommand({ logGroupName: groupName }),
            );
            assert.notStrictEqual(
              resp.logStreams?.some((s) => s.logStreamName, "stream-1"),
              "DeleteLogStream: stream-1 still present after delete",
            );
          },
        },
      ],
      setup: async (ctx) => {
        const { logs } = makeClients(ctx);
        const groupName = `/overcast/${ctx.runId}/events`;
        await logs.send(new CreateLogGroupCommand({ logGroupName: groupName }));
        await logs.send(
          new CreateLogStreamCommand({
            logGroupName: groupName,
            logStreamName: "stream-1",
          }),
        );
      },
      teardown: async (ctx) => {
        const { logs } = makeClients(ctx);
        const groupName = `/overcast/${ctx.runId}/events`;
        try {
          await logs.send(
            new DeleteLogGroupCommand({ logGroupName: groupName }),
          );
        } catch {}
      },
    },
  ];
}

File: compat/suites/python-sdk/groups/cloudwatch_logs.py
"""
groups/cloudwatch_logs.py — CloudWatch Logs compatibility test implementations.
"""

from __future__ import annotations
import time
from lib.harness import TestContext
from lib.clients import make_clients


def _logs(ctx: TestContext):
    return make_clients(ctx.endpoint, ctx.region).logs


# ── logs-groups ───────────────────────────────────────────────────────────────

def setup_logs_groups(ctx: TestContext) -> None:
    logs = _logs(ctx)
    name = f"/compat/{ctx.run_id}"
    logs.create_log_group(logGroupName=name)
    ctx["log_group"] = name


def teardown_logs_groups(ctx: TestContext) -> None:
    name = ctx.get("log_group")
    if name:
        try:
            _logs(ctx).delete_log_group(logGroupName=name)
        except Exception:
            pass


def CreateLogGroup(ctx: TestContext) -> None:
    logs = _logs(ctx)
    name = f"/compat/{ctx.run_id}-create"
    logs.create_log_group(logGroupName=name)
    try:
        resp = logs.describe_log_groups(logGroupNamePrefix=name)
        groups = resp.get("logGroups", [])
        if not any(g["logGroupName"] == name for g in groups):
            raise AssertionError(f"CreateLogGroup: {name!r} not found after creation")
    finally:
        logs.delete_log_group(logGroupName=name)


def DescribeLogGroups(ctx: TestContext) -> None:
    logs = _logs(ctx)
    name = ctx["log_group"]
    resp = logs.describe_log_groups(logGroupNamePrefix=name)
    groups = resp.get("logGroups", [])
    if not any(g["logGroupName"] == name for g in groups):
        raise AssertionError(f"DescribeLogGroups: {name!r} not found")


def PutRetentionPolicy(ctx: TestContext) -> None:
    logs = _logs(ctx)
    name = ctx["log_group"]
    logs.put_retention_policy(logGroupName=name, retentionInDays=7)
    resp = logs.describe_log_groups(logGroupNamePrefix=name)
    groups = [g for g in resp.get("logGroups", []) if g["logGroupName"] == name]
    assert groups and groups[0].get("retentionInDays") == 7, "PutRetentionPolicy: retention!=7"


def VerifyRetentionPolicy(ctx: TestContext) -> None:
    logs = _logs(ctx)
    name = ctx["log_group"]
    resp = logs.describe_log_groups(logGroupNamePrefix=name)
    groups = resp.get("logGroups", [])
    matching = [g for g in groups if g["logGroupName"] == name]
    if not matching:
        raise AssertionError(f"VerifyRetentionPolicy: log group {name!r} not found")
    if matching[0].get("retentionInDays") != 7:
        raise AssertionError(f"VerifyRetentionPolicy: expected 7 days, got {matching[0].get('retentionInDays')}")


def DeleteRetentionPolicy(ctx: TestContext) -> None:
    logs = _logs(ctx)
    name = ctx["log_group"]
    logs.delete_retention_policy(logGroupName=name)
    resp = logs.describe_log_groups(logGroupNamePrefix=name)
    groups = [g for g in resp.get("logGroups", []) if g["logGroupName"] == name]
    assert groups and "retentionInDays" not in groups[0], "DeleteRetentionPolicy: retention still set"


def DeleteLogGroup(ctx: TestContext) -> None:
    logs = _logs(ctx)
    name = f"/compat/{ctx.run_id}-del"
    logs.create_log_group(logGroupName=name)
    logs.delete_log_group(logGroupName=name)
    resp = logs.describe_log_groups(logGroupNamePrefix=name)
    groups = resp.get("logGroups", [])
    if any(g["logGroupName"] == name for g in groups):
        raise AssertionError(f"DeleteLogGroup: {name!r} still listed after deletion")


def CreateLogStream(ctx: TestContext) -> None:
    logs = _logs(ctx)
    name = ctx["log_group"]
    stream = f"stream-grp-{ctx.run_id}"
    logs.create_log_stream(logGroupName=name, logStreamName=stream)


def TagLogGroup(ctx: TestContext) -> None:
    logs = _logs(ctx)
    name = ctx["log_group"]
    logs.tag_log_group(logGroupName=name, tags={"env": "test"})


# ── logs-events ───────────────────────────────────────────────────────────────

def setup_logs_events(ctx: TestContext) -> None:
    logs = _logs(ctx)
    name = f"/compat/{ctx.run_id}-events"
    stream = "mystream"
    logs.create_log_group(logGroupName=name)
    logs.create_log_stream(logGroupName=name, logStreamName=stream)
    ctx["log_events_group"] = name
    ctx["log_events_stream"] = stream


def teardown_logs_events(ctx: TestContext) -> None:
    name = ctx.get("log_events_group")
    if name:
        try:
            _logs(ctx).delete_log_group(logGroupName=name)
        except Exception:
            pass


def PutLogEvents(ctx: TestContext) -> None:
    logs = _logs(ctx)
    name = ctx["log_events_group"]
    stream = ctx["log_events_stream"]
    now_ms = int(time.time() * 1000)
    logs.put_log_events(
        logGroupName=name,
        logStreamName=stream,
        logEvents=[
            {"timestamp": now_ms, "message": "event one"},
            {"timestamp": now_ms + 1, "message": "event two"},
        ],
    )


def GetLogEvents(ctx: TestContext) -> None:
    logs = _logs(ctx)
    name = ctx["log_events_group"]
    stream = ctx["log_events_stream"]
    resp = logs.get_log_events(
        logGroupName=name,
        logStreamName=stream,
        startFromHead=True,
    )
    events = resp.get("events", [])
    if not events:
        raise AssertionError("GetLogEvents: no events returned")
    messages = [e["message"] for e in events]
    if "event one" not in messages:
        raise AssertionError(f"GetLogEvents: expected 'event one' in {messages}")


def FilterLogEvents(ctx: TestContext) -> None:
    logs = _logs(ctx)
    name = ctx["log_events_group"]
    resp = logs.filter_log_events(logGroupName=name, filterPattern="event two")
    events = resp.get("events", [])
    if not events:
        raise AssertionError("FilterLogEvents: no events returned matching 'event two'")


def DescribeLogStreams(ctx: TestContext) -> None:
    logs = _logs(ctx)
    name = ctx["log_events_group"]
    stream = ctx["log_events_stream"]
    resp = logs.describe_log_streams(logGroupName=name)
    streams = [s["logStreamName"] for s in resp.get("logStreams", [])]
    if stream not in streams:
        raise AssertionError(f"DescribeLogStreams: {stream!r} not found in {streams}")


def DeleteLogStream(ctx: TestContext) -> None:
    logs = _logs(ctx)
    name = ctx["log_events_group"]
    stream = ctx["log_events_stream"]
    logs.delete_log_stream(logGroupName=name, logStreamName=stream)
    resp = logs.describe_log_streams(logGroupName=name)
    streams = [s["logStreamName"] for s in resp.get("logStreams", [])]
    if stream in streams:
        raise AssertionError(f"DeleteLogStream: {stream!r} still listed")


# ── ImplMap ───────────────────────────────────────────────────────────────────

IMPLS = {
    "CreateLogGroup": CreateLogGroup,
    "DescribeLogGroups": DescribeLogGroups,
    "PutRetentionPolicy": PutRetentionPolicy,
    "VerifyRetentionPolicy": VerifyRetentionPolicy,
    "DeleteRetentionPolicy": DeleteRetentionPolicy,
    "CreateLogStream": CreateLogStream,
    "TagLogGroup": TagLogGroup,
    "DeleteLogGroup": DeleteLogGroup,
    "PutLogEvents": PutLogEvents,
    "GetLogEvents": GetLogEvents,
    "FilterLogEvents": FilterLogEvents,
    "DescribeLogStreams": DescribeLogStreams,
    "DeleteLogStream": DeleteLogStream,
}

SETUP = {
    "logs-groups": setup_logs_groups,
    "logs-events": setup_logs_events,
}

TEARDOWN = {
    "logs-groups": teardown_logs_groups,
    "logs-events": teardown_logs_events,
}
